addons/web/models/ir_actions_server.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from uuid import uuid4
+from json import JSONDecodeError
+from secrets import compare_digest
+import hashlib
+
+from odoo.http import request
+from odoo import fields, models, api, exceptions, _
+
+
+class ServerAction(models.Model):
+    _inherit = "ir.actions.server"
+
+    url = fields.Char("Public Url", compute="_compute_url")
+    url_path = fields.Char("URL Path", help="Additional URL path", compute="_compute_url_path", store=True)
+    expose = fields.Selection([("webhook", "Expose as webhook")],
+        help="Expose this server action via a controller. The 'as webhook' option will make the server action available via '/web/hook/<url_path>'")
+
+    log_webhook_calls = fields.Boolean(string="Log Calls", default=True)
+
+    @api.constrains("expose", "state")
+    def _constrains_state(self):
+        for action in self:
+            if action.expose == "webhook" and action.state != "code":
+                raise exceptions.ValidationError(_("A webhook is necessarily defined as executing python code"))
+
+    @api.depends("state", "expose", "url_path")
+    def _compute_url(self):
+        for action in self:
+            if action.expose == "webhook" and action.state == "code":
+                action.url = f"/web/hook/{action.url_path}"
+            else:
+                action.url = False
+
+    @api.depends("state", "expose")
+    def _compute_url_path(self):
+        for action in self:
+            if action.expose == "webhook" and action.state == "code":
+                action.url_path = str(uuid4())
+            else:
+                action.url_path = False
+
+    def _get_eval_context(self, action=None):
+        eval_context = super()._get_eval_context(action)
+        if action and action.expose == "webhook" and request:
+            try:
+                payload = request.get_json_data()
+            except JSONDecodeError:
+                payload = request.get_http_params()
+
+            def _compare_digest(a, b):
+                if not compare_digest(a, b):
+                    raise exceptions.UserError(_("Secrets not matching"))
+            eval_context.update({
+                "request": request,
```

Comment:
website also add request (and json), we could probably add request if action.expose != false.

Do we really need payload if we have request?

Reply:
i think it's simpler (even for technical users) to know that they have a `payload` dict rather than having to dive into werkzeug object attributes (which depends on whether you're doing POST or GET) or into specific http odoo stuff.